xds/internal/server/rds_handler.go

```diff
@@ -89,35 +95,37 @@ func (rh *rdsHandler) updateRouteNamesToWatch(routeNamesToWatch map[string]bool)
 			delete(rh.updates, routeName)
 		}
 	}
-
-	// If the full list (determined by length) of updates are now successfully
-	// updated, the listener is ready to be updated.
-	if len(rh.updates) == len(rh.cancels) && len(routeNamesToWatch) != 0 {
-		drainAndPush(rh.updateChannel, rdsHandlerUpdate{updates: rh.updates})
-	}
 }
 
-// handleRouteUpdate persists the route config for a given route name, and also
-// sends an update to the Listener Wrapper on an error received or if the rds
-// handler has a full collection of updates.
-func (rh *rdsHandler) handleRouteUpdate(routeName string, update xdsresource.RouteConfigUpdate) {
-	rh.mu.Lock()
-	defer rh.mu.Unlock()
-	rh.updates[routeName] = update
+// determines if all dynamic RouteConfiguration needed has received
+// configuration or update. Must be called from an xDS Client Callback.
+func (rh *rdsHandler) determineRouteConfigurationReady() bool {
+	// Safe to read cancels because only written to in other parts of xDS Client
+	// Callbacks, which are sync.
+	return len(rh.updates) == len(rh.cancels)
+}
 
-	// If the full list (determined by length) of updates have successfully
-	// updated, the listener is ready to be updated.
-	if len(rh.updates) == len(rh.cancels) {
-		drainAndPush(rh.updateChannel, rdsHandlerUpdate{updates: rh.updates})
+// Must be called from an xDS Client Callback.
+func (rh *rdsHandler) handleRouteUpdate(routeName string, update rdsWatcherUpdate) {
+	rwu, ok := rh.updates[routeName]
+	if !ok {
+		rwu = rdsWatcherUpdate{}
 	}
-}
 
-func drainAndPush(ch chan rdsHandlerUpdate, update rdsHandlerUpdate) {
-	select {
-	case <-ch:
-	default:
+	if update.err != nil {
+		if xdsresource.ErrType(update.err) == xdsresource.ErrorTypeResourceNotFound {
+			// Clear update. This will cause future RPCs on connections which
+			// use this route configuration to fail with UNAVAILABLE.
+			rwu.update = nil
+		}
+		// Write error.
+		rwu.err = update.err
```

Comment:
IIUC this results in both `rwu.update` set and `rwu.err` set in cases where the error is not `NotFound` - is that right?

Reply:
Yeah it's intended because update is used when it's set irrespective of error, so this is WAI. However Easwar raised a point to which I tried that is the same vein as this: https://github.com/grpc/grpc-go/pull/6889#discussion_r1436596753. Since both of you brought it up I can try and change it.

Reply:
Wait actually I think this is an orthogonal point, but I do think this is right, but the writing of error is sort of a no op. Functionally, I want to ignore errors (not resource not found) after good updates, so I write the error but leave the update and the update gets used. In this case the error will never be read, but I think it's fine to leave as is since on the first error you need to write it.

Reply:
Can this whole if/else block be expressed as:

```go
// Accept the new update if any of the following are true:
// 1. we had no valid update data
// 2. the update is valid
// 3. the update error is ResourceNotFOund
if rwu.update == nil || update.err == nil || ErrType == ResourceNotFound {
	rwu = update
}
```

It's also a bit confusing that `update` has a field called `update`.  Can we call it `data` or something instead?

Reply:
Good idea. Switched to your solution, switched name to data.